### PR TITLE
8336702: C2 compilation fails with "all memory state should have been processed" assert

### PR DESCRIPTION
**Pull request by rwestrel**
When converting a `LongCountedLoop` into a loop nest, c2 needs jvm
state to add predicates to the inner loop. For that, it peels an
iteration of the loop and uses the state of the safepoint at the end
of the loop. That's only legal if there's no side effect between the
safepoint and the backedge that goes back into the loop. The assert
failure here happens in code that checks that.

That code compares the memory states at the safepoint and at the
backedge. If they are the same then there's no side effect. To check
consistency, the `MergeMem` at the safepoint is cloned. As the logic
iterates over the backedge state, it clears every component of the
state it encounters from the `MergeMem`. Once done, the cloned
`MergeMem` should be "empty". In the case of this failure, no side
effect is found but the cloned `MergeMem` is not empty. That happens
because of EA: it adds edges to the `MergeMem` at the safepoint that
it doesn't add to the backedge `Phis`.

So it's the verification code that fails and I propose dealing with
this by ignoring memory state added by EA in the verification code.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8336702](https://bugs.openjdk.org/browse/JDK-8336702): C2 compilation fails with "all memory state should have been processed" assert (**Bug** - P3)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21009/head:pull/21009` \
`$ git checkout pull/21009`

Update a local copy of the PR: \
`$ git checkout pull/21009` \
`$ git pull https://git.openjdk.org/jdk.git pull/21009/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21009`

View PR using the GUI difftool: \
`$ git pr show -t 21009`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21009.diff">https://git.openjdk.org/jdk/pull/21009.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21009#issuecomment-2352318142)

### CODE REVIEW

File: test/hotspot/jtreg/compiler/longcountedloops/TestSafePointWithEAState.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  public static void main(String[] m) {
+    TestSafePointWithEAState o = new TestSafePointWithEAState();
+    o.c();
+  }
+}
+
+class h {
+  float n;
+  h(float n) { this.n = n; }
+}
```
Comment (eme64) [inline test/hotspot/jtreg/compiler/longcountedloops/TestSafePointWithEAState.java:59]: Java indentation is supposed to be 4 spaces ;)
Adding some explicit brackets would also be nice, but that is more subjective.

**Review by eme64 — CHANGES_REQUESTED** [review]
Drive-by comment. I don't understand enough about EA to review the VM part.

File: src/hotspot/share/opto/loopnode.cpp [hunk 2]
```diff
@@ -692,14 +692,24 @@ SafePointNode* PhaseIdealLoop::find_safepoint(Node* back_control, Node* x, Ideal
...
+    // reset.
     MergeMemNode* mm = nullptr;
 #ifdef ASSERT
     if (mem->is_MergeMem()) {
       mm = mem->clone()->as_MergeMem();
       _igvn._worklist.push(mm);
       for (MergeMemStream mms(mem->as_MergeMem()); mms.next_non_empty(); ) {
-        if (mms.alias_idx() != Compile::AliasIdxBot && loop != get_loop(ctrl_or_self(mms.memory()))) {
+        // Loop invariant memory state won't be reset by no_side_effect_since_safepoint(). Do it here.
+        // Escape Analysis can add state to mm that it doesn't add to the backedge memory Phis, breaking verification
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopnode.cpp:708]: Where exactly does that happen in EA?

File: test/hotspot/jtreg/compiler/longcountedloops/TestSafePointWithEAState.java (hunk 1 above)
Comment (rwestrel) [reply test/hotspot/jtreg/compiler/longcountedloops/TestSafePointWithEAState.java]: Right. Fixed in new commit.

File: src/hotspot/share/opto/loopnode.cpp (hunk 2 above)
Comment (TobiHartmann) [reply src/hotspot/share/opto/loopnode.cpp:708]: Okay, thanks for the details!

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. Testing passed.

**Comment by rwestrel** [comment]
Anyone else for a review?

**Review by chhagedorn — APPROVED** [review]
Looks good to me, too!

**Comment by rwestrel** [comment]
@TobiHartmann @chhagedorn thanks for the reviews.

**Pull request closed without merging**